src/unitxt/metrics.py

```diff
@@ -400,31 +699,180 @@ def process(self, stream: Stream, stream_name: Optional[str] = None) -> Generato
+                        f"Missing '{cc}' from instance {instance}. "
+                        f"This field is required for group based metric computation."
+                    )
+            group_key = additional_inputs[self.grouping_field]
+            # indicator if is in the baseline group or not
+            is_baseline = str(additional_inputs[subgroup_field]) == baseline_name
+            # convert True (baseline) to 0, and False (others) to 1, store in respective groups
+            idx = int(not is_baseline)
+
+            for field_name in field_names:
+                group_to_instance_scores[group_key][field_name][idx].append(
+                    instance["score"]["instance"][field_name]
+                )
+
+        # now for each group, take the aggregation function, comparing others to baseline
+        return [
+            {
+                "score": {
+                    "instance": {
+                        field_name: group_aggregation_func(
+                            baseline_scores=scores[field_name][0],
+                            other_scores=scores[field_name][1],
+                        )
+                        for field_name in field_names
+                    }
+                }
+            }
+            for scores in group_to_instance_scores.values()
+        ]
+
+    def _set_up_group_mean_aggregation(
+        self, instances, reduction_params, reduction_fields
+    ):
+        # if treat groups as units
+        if reduction_params["agg_func"][2]:
+            # pass the group aggregate---not instance---scores to resample as usual
+            aggregation_function = self.average_instance_scores
+            instances_to_resample = self.get_group_aggregated_instance_scores(
+                instances, reduction_fields, reduction_params["agg_func"][1]
+            )
+        else:
+            # pass the instance scores to resample, and calculate the group aggregation on the resamplings
+            instances_to_resample = deepcopy(instances)
+
+            def aggregation_function(
+                instances,
+                field_name,
+                group_aggregation_func=reduction_params["agg_func"][1],
+            ):
+                group_scores = self.get_group_aggregated_instance_scores(
+                    instances, [field_name], group_aggregation_func
+                )
+                return nan_mean(
+                    [group["score"]["instance"][field_name] for group in group_scores]
+                )
+
+        return instances_to_resample, aggregation_function
+
+    def _set_up_group_mean_subgroup_comparison_aggregation(
+        self, instances, reduction_params, reduction_fields
+    ):
+        # if treat groups as units
+        if reduction_params["agg_func"][2]:
+            # pass the group aggregate---not instance---scores to resample as usual
+            aggregation_function = self.average_instance_scores
+            instances_to_resample = (
```

Comment:
IIUC these are now the `scores_to_resample` right?
since the computation per group is done in `get_group_aggregated_instance_scores_by_subgroups`?

Reply:
yes: in the _set_up functions, we prepare the inputs to the bootstrapping.  If grouping is requested (reduction_params["agg_func"][2] is True), then the scores are first calculated on the groups, at which point you need to simply average them.  If not, you need to pass the original instance scores (hence the deepcopy) to resampling, and THEN apply the group aggregation function after resampling each time.